pandas/plotting/_matplotlib/core.py

```diff
@@ -236,6 +253,54 @@ def _validate_color_args(self):
                             "pass 'style' without a color symbol"
                         )
 
+    @staticmethod
+    def _create_iter_data_given_by(
+        data: DataFrame, by: Optional[List]
+    ) -> Union[DataFrame, Dict[str, Union[DataFrame, Series]]]:
+        """
+        Create data for iteration given `by` is assigned or not, and it is only
+        used in both hist and boxplot.
+
+        If `by` is assigned, return a dictionary of DataFrames in which the key of
+        dictionary is the values in groups.
+        If `by` is not assigned, return input as is, and this preserves current
+        status of iter_data.
+
+        Parameters
+        ----------
+        data: reformatted grouped data from `_compute_plot_data` method
+        by: list or None, value assigned to `by`.
+
+        Returns
+        -------
+        iter_data: DataFrame or Dictionary of DataFrames
+
+        Examples
+        --------
+        If `by` is assigned:
+
+        >>> tuples = [('h1', 'a'), ('h1', 'b'), ('h2', 'a'), ('h2', 'b')]
+        >>> mi = MultiIndex.from_tuples(tuples)
+        >>> value = [[1, 3, np.nan, np.nan],
+        ...          [3, 4, np.nan, np.nan], [np.nan, np.nan, 5, 6]]
+        >>> data = DataFrame(value, columns=mi)
+        >>> _create_iter_data_given_by(data, by=["col1"]) # doctest: +SKIP
+        {'h1': DataFrame({'a': [1, 3, np.nan], 'b': [3, 4, np.nan]}),
+         'h2': DataFrame({'a': [np.nan, np.nan, 5], 'b': [np.nan, np.nan, 6]})}
+        """
+        iter_data: Union[DataFrame, Dict[str, Union[DataFrame, Series]]]
+        if not by:
+            iter_data = data
+        else:
+            # Select sub-columns based on the value of first level of MI
+            # TODO: mypy complains because Index does not have levels, only MI has.
```

Comment:
Rather than type ignore can you add an assertion that `data.columns` is a MultiIndex?

Reply:
nice! changed!